drivers/retained_mem/retained_mem_zephyr_ram.c

```diff
@@ -0,0 +1,149 @@
+
+	k_mutex_lock(&data->lock, K_FOREVER);
+#endif
+
+	memcpy(buffer, (void *)(config->address + offset), size);
+
+#ifdef CONFIG_MULTITHREADING
+	k_mutex_unlock(&data->lock);
+#endif
+
+	return 0;
+}
+
+static int32_t zephyr_retained_mem_ram_write(const struct device *dev, size_t offset,
+					     const uint8_t *buffer, size_t size)
+{
+	const struct zephyr_retained_mem_ram_config *config = dev->config;
+
+#ifdef CONFIG_MULTITHREADING
+	struct zephyr_retained_mem_ram_data *data = dev->data;
+
+	k_mutex_lock(&data->lock, K_FOREVER);
+#endif
+
+	memcpy((void *)(config->address + offset), buffer, size);
+
+#ifdef CONFIG_MULTITHREADING
+	k_mutex_unlock(&data->lock);
+#endif
+
+	return 0;
+}
+
+static int32_t zephyr_retained_mem_ram_clear(const struct device *dev)
+{
+	const struct zephyr_retained_mem_ram_config *config = dev->config;
+
+#ifdef CONFIG_MULTITHREADING
+	struct zephyr_retained_mem_ram_data *data = dev->data;
+
+	k_mutex_lock(&data->lock, K_FOREVER);
+#endif
+
+	memset((void *)config->address, 0, config->size);
+
+#ifdef CONFIG_MULTITHREADING
+	k_mutex_unlock(&data->lock);
+#endif
+
+	return 0;
+}
+
+static const struct retained_mem_driver_api zephyr_retained_mem_ram_api = {
+	.size = zephyr_retained_mem_ram_size,
+	.read = zephyr_retained_mem_ram_read,
+	.write = zephyr_retained_mem_ram_write,
+	.clear = zephyr_retained_mem_ram_clear,
+};
+
+#define ZEPHYR_RETAINED_MEM_RAM_DEVICE(inst)							\
+	static const struct zephyr_retained_mem_ram_config					\
+			zephyr_retained_mem_ram_config_##inst = {				\
+		.address = DT_REG_ADDR(DT_PARENT(DT_INST(inst, DT_DRV_COMPAT))),		\
+		.size = DT_REG_SIZE(DT_PARENT(DT_INST(inst, DT_DRV_COMPAT))),			\
+	};											\
+	DEVICE_DT_INST_DEFINE(inst,								\
+			      &zephyr_retained_mem_ram_init,					\
+			      NULL,								\
+			      NULL,								\
+			      &zephyr_retained_mem_ram_config_##inst,				\
+			      POST_KERNEL,							\
+			      CONFIG_RETAINED_MEM_INIT_PRIORITY,				\
+			      &zephyr_retained_mem_ram_api);
+
+#define ZEPHYR_RETAINED_MEM_RAM_DEVICE_MULTITHREADING(inst)					\
+	static struct zephyr_retained_mem_ram_data zephyr_retained_mem_ram_data_##inst;		\
+	static const struct zephyr_retained_mem_ram_config					\
+			zephyr_retained_mem_ram_config_##inst = {				\
+		.address = DT_REG_ADDR(DT_PARENT(DT_INST(inst, DT_DRV_COMPAT))),		\
+		.size = DT_REG_SIZE(DT_PARENT(DT_INST(inst, DT_DRV_COMPAT))),			\
+	};											\
+	DEVICE_DT_INST_DEFINE(inst,								\
+			      &zephyr_retained_mem_ram_init,					\
+			      NULL,								\
+			      &zephyr_retained_mem_ram_data_##inst,				\
+			      &zephyr_retained_mem_ram_config_##inst,				\
+			      POST_KERNEL,							\
+			      CONFIG_RETAINED_MEM_INIT_PRIORITY,				\
+			      &zephyr_retained_mem_ram_api);
+
+#ifdef CONFIG_MULTITHREADING
+DT_INST_FOREACH_STATUS_OKAY(ZEPHYR_RETAINED_MEM_RAM_DEVICE_MULTITHREADING)
+#else
+DT_INST_FOREACH_STATUS_OKAY(ZEPHYR_RETAINED_MEM_RAM_DEVICE)
+#endif
```

Comment:
Similar nonblocking suggestion for how to simplify this applies here too

Reply:
Fixed